wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java

```diff
@@ -154,24 +155,40 @@
   });
 
   private final Schema schema;
+  private final ImmutableMap<ProtoType, TypeName> nameToWireJavaName;
   private final ImmutableMap<ProtoType, TypeName> nameToJavaName;
+  private final ImmutableMap<ProtoType, TypeName> protoTypeToAdapter;
   private final boolean emitAndroid;
   private final boolean emitCompact;
 
-  private JavaGenerator(Schema schema, ImmutableMap<ProtoType, TypeName> nameToJavaName,
+  private JavaGenerator(Schema schema, ImmutableMap<ProtoType,
+      TypeName> nameToWireJavaName, ImmutableMap<ProtoType, TypeName> nameToJavaName,
+      ImmutableMap<ProtoType, TypeName> protoTypeToAdapter,
       boolean emitAndroid, boolean emitCompact) {
     this.schema = schema;
+    this.nameToWireJavaName = nameToWireJavaName;
     this.nameToJavaName = nameToJavaName;
+    this.protoTypeToAdapter = protoTypeToAdapter;
     this.emitAndroid = emitAndroid;
     this.emitCompact = emitCompact;
   }
 
   public JavaGenerator withAndroid(boolean emitAndroid) {
-    return new JavaGenerator(schema, nameToJavaName, emitAndroid, emitCompact);
+    return new JavaGenerator(schema, nameToWireJavaName, nameToJavaName, protoTypeToAdapter,
+        emitAndroid, emitCompact);
   }
 
   public JavaGenerator withCompact(boolean compactGeneration) {
-    return new JavaGenerator(schema, nameToJavaName, emitAndroid, compactGeneration);
+    return new JavaGenerator(schema, nameToWireJavaName, nameToJavaName, protoTypeToAdapter,
+        emitAndroid, compactGeneration);
+  }
+
+  public JavaGenerator withCustomProtoAdapter(Map<ProtoType, TypeName> nameToJavaName,
```

Comment:
* Name should be plural since you're accepting multiple
* Should we use Guava's `Table` here? @swankjesse might have opinions.
* I think we probably want to use `ClassName` here just so no one does anything stupid and passes like a generic or primitive or something.